components/support/images/src/main/java/mozilla/components/support/images/compose/loader/ImageLoader.kt

```diff
@@ -0,0 +1,178 @@
+    targetSize: Dp = defaultTargetSize,
+    minSize: Dp = targetSize / DEFAULT_MIN_MAX_MULTIPLIER,
+    maxSize: Dp = targetSize * DEFAULT_MIN_MAX_MULTIPLIER,
+    maxScaleFactor: Float = DEFAULT_MAXIMUM_SCALE_FACTOR,
+    content: @Composable ImageLoaderScope.() -> Unit
+) {
+    val desiredSize = with(LocalDensity.current) {
+        DesiredSize(
+            targetSize = targetSize.roundToPx(),
+            minSize = minSize.roundToPx(),
+            maxSize = maxSize.roundToPx(),
+            maxScaleFactor = maxScaleFactor
+        )
+    }
+
+    val scope = remember(url) {
+        InternalImageLoaderScope(
+            client,
+            url,
+            private,
+            connectTimeout,
+            readTimeout,
+            desiredSize
+        )
+    }
+
+    if (!scope.hasStartedLoading.value) {
+        scope.PerformLoad()
+    }
+
+    scope.content()
+}
+
+@Composable
+private fun InternalImageLoaderScope.PerformLoad() {
+    hasStartedLoading.value = true
+
+    LaunchedEffect(url) {
+        val bitmap = withContext(Dispatchers.IO) {
+            fetchAndDecode(
+                client,
+                url,
+                private,
+                connectTimeout,
+                readTimeout,
+                desiredSize
+            )
+        }
+
+        if (bitmap != null) {
+            loaderState.value = ImageLoaderState.Image(BitmapPainter(bitmap.asImageBitmap()))
+        } else {
+            loaderState.value = ImageLoaderState.Failed
+        }
+    }
+}
+
+@Suppress("LongParameterList")
+private suspend fun fetchAndDecode(
+    client: Client,
+    url: String,
+    private: Boolean,
+    connectTimeout: Pair<Long, TimeUnit>,
+    readTimeout: Pair<Long, TimeUnit>,
+    desiredSize: DesiredSize
+): Bitmap? = withContext(Dispatchers.IO) {
+    val data = fetch(
+        client,
+        url,
+        private,
+        connectTimeout,
+        readTimeout
+    ) ?: return@withContext null
+
+    decoders.firstNotNullOfOrNull { decoder ->
+        decoder.decode(data, desiredSize)
+    }
+}
+
+private fun fetch(
+    client: Client,
+    url: String,
+    private: Boolean,
+    connectTimeout: Pair<Long, TimeUnit>,
+    readTimeout: Pair<Long, TimeUnit>,
+): ByteArray? {
+    val request = Request(
+        url = url.trim(),
+        method = Request.Method.GET,
+        cookiePolicy = if (private) {
+            Request.CookiePolicy.OMIT
+        } else {
+            Request.CookiePolicy.INCLUDE
+        },
+        connectTimeout = connectTimeout,
+        readTimeout = readTimeout,
+        redirect = Request.Redirect.FOLLOW,
+        useCaches = true,
+        private = private
+    )
+
+    return try {
+        val response = client.fetch(request)
+        if (response.isSuccess) {
+            response.body.useStream { it.readBytes() }
+        } else {
+            null
+        }
+    } catch (e: IOException) {
+        null
+    }
+}
+
+private const val DEFAULT_CONNECT_TIMEOUT = 2L // Seconds
```

Comment:
nit: should these constants be moved to the top?

Reply:
That's a great question. I don't think we have any specific rules about them. My personal kind of thinking is: If they are public and important to see/know then I put them at the top. If they are constants that are not really important to know or even distracting then I move them to the bottom.

Reply:
I have seen companion objects placed at the top, and that is wrong, IMO and I think also against official Kotlin coding conventions. If constants are not part of a companion object, my preference is the same... Relevant ones should be placed at the top, but constants declared just to avoid Magic Number warning do not need increased visibility.